[19.03 backport] various dockerfile changes and update to buster variant

thaJeztah opened the pull request:
This backports various changes from master to the 19.03 branch, some of which are in preparation of updating to Go 1.13, others to reduce formatting differences to make comparing slightly easier.

Note that in this variant, I did not include the Dockerfile changes to the "buildkit + buildx optimized" version (https://github.com/moby/moby/pull/39340); as a result, some (formatting) changes did not apply cleanly.

If we think the buildkit optimized version should be considered, I can open a variant with that PR (and related/follow-up PRs) included.

- https://github.com/moby/moby/commit/961119db21b95504f819a31dfadd7115757fffb3 Dockerfile: set GO111MODULE=off
    - taken from https://github.com/moby/moby/pull/39549 Bump go 1.13.0
- https://github.com/moby/moby/pull/40038 go-swagger: fix panic under Golang 1.13
    - no clean cherry-pick, because https://github.com/moby/moby/pull/39340 is not in the 19.03 branch
- https://github.com/moby/moby/pull/39919 Use build args to override binary commits in dockerfile
- https://github.com/moby/moby/pull/40359 bump vndr v0.1.0 to support versioned import paths
- https://github.com/moby/moby/pull/40471 bump vndr v0.1.1
- https://github.com/moby/moby/pull/40044 Dockerfile formatting changes and optimisations
    - no clean cherry-pick, because https://github.com/moby/moby/pull/39340 is not in the 19.03 branch
- https://github.com/moby/moby/pull/39880 Dockerfile: switch golang image to "buster" variant
- https://github.com/moby/moby/pull/40526 Update windows container utility to fix compilation on gcc-mingw-w64 8.3

thaJeztah committed:
Dockerfile: switch golang image to "buster" variant, and update btrfs packages

The btrfs-tools was a transitional package, and no longer exists:

> Package btrfs-tools
> stretch (oldstable) (admin): transitional dummy package
> 4.7.3-1: amd64 arm64 armel armhf i386 mips mips64el mipsel ppc64el s390x

It must be replaced either by `btrfs-progs` or `libbtrfs-dev` (which has just the development headers)

> Package: libbtrfs-dev (4.20.1-2)
> Checksumming Copy on Write Filesystem utilities (development headers)

Note that the `libbtrfs-dev` package is not available on Debian stretch
(only in stretch-backports)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit 4e3ab9e9fbca682f75eb350c8ad4312282869a03)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Dockerfile: switch to iptables-legacy to match the host

CI runs on Ubuntu 16.04 machines, which use iptables (legacy), but
Debian buster uses nftables. Because of this, DNS resolution does not
work if the daemon configures iptables.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit bb0472bd23806327a1ff1ea87343a6a0cadc21e9)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Dockerfile: set GO111MODULE=off

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit 961119db21b95504f819a31dfadd7115757fffb3)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

jmartin84 committed:
Use build args to override binary commits in dockerfile

Signed-off-by: Justen Martin <jmart@the-coder.com>
(cherry picked from commit 095ca77f4810b4e074ee47894f68434fece612c5)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

kolyshkin committed:
go-swagger: fix panic

This is an attempt to fix go-swagger panic under Golang 1.13.

Details:
 * https://github.com/go-openapi/jsonpointer/pull/4
 * https://github.com/go-swagger/go-swagger/pull/2059

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>
(cherry picked from commit 93f9b902af89f82367d750aa871d40f25ccd99ca)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Dockerfile: use spaces for indentation

Indenting with tabs can cause the formatting to go wonky,
because the first line of any command is "indented" with spaces,
but following lines are not, therefore they can be mis-aligned with
the first line.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit a42b4144bce728590d7c01aa2401589e71850518)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Dockerfile: use build-arg for vpnkit

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit 1cfcce5e210ad7a0296fe341b3dda40a0af06c12)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Dockerfile: sort packages alphabetically

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit ee0ef6c535dc8a3ab62266d7da7801dd7c5b4ea8)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Dockerfile: order COPY lines by change frequency

Ordering the COPY lines to optimize for layer sharing
when these dependencies are updated.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit 8edbe5dec254d19e378a5a7cb1abf202f7b20eef)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Dockerfile: align consecutive COPY lines

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit 93edf327dc6597ee1a763bc105ba7e341415edd8)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
hack/make: ignore failure to stop apparmor

```
 ---> Making bundle: .integration-daemon-stop (in bundles/test-integration)
 ++++ cat bundles/test-integration/docker.pid
 +++ kill 13137
 +++ /etc/init.d/apparmor stop
 Leaving: AppArmorNo profiles have been unloaded.

 Unloading profiles will leave already running processes permanently
 unconfined, which can lead to unexpected situations.

 To set a process to complain mode, use the command line tool
 'aa-complain'. To really tear down all profiles, run 'aa-teardown'."

script returned exit code 255
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit 5dbfae694919b1aff76f360061b4065349bf1900)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
bump vndr v0.1.0 to support versioned import paths

With this change, go packages/modules that use versioned
import paths (github.com/foo/bar/v2), but don't use a directory
in the repository, can now be supported.

For example:

```
github.com/coreos/go-systemd/v22 v22.0.0
```

will vendor the github.com/coreos/go-systemd repository
into `vendor/github.com/coreos/go-systemd/v22`.

full diff: https://github.com/LK4D4/vndr/compare/f5ab8fc5fb64d66b5c6e55a0bcb58b2e92362fa0...v0.1.0

- LK4D4/vndr#83 migrate bitbucket to api 2.0
    - fixes LK4D4/vndr#82 https://api.bitbucket.org/1.0/repositories/ww/goautoneg: 410 Gone
- LK4D4/vndr#86 Replace sort.Sort with sort.Strings
- LK4D4/vndr#87 support `github.com/coreos/go-systemd/v22`

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit d4f05c168d4497944a1427d830bdf9e69fb2f4c7)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
bump vndr v0.1.1

full diff: https:/github.com/LK4D4/vndr/compare/v0.1.0...v0.1.1

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit 486161a63a841fae4f70558d1204db483006a017)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
bump windows-container-utility aa1ba87e99b68e0113bd27ec26c60b88f9d4ccd9

full diff: https://github.com/docker/windows-container-utility/compare/e004a1415a433447369e315b9d7df357102be0d2...aa1ba87e99b68e0113bd27ec26c60b88f9d4ccd9

changes:

- Use standard include paths instead of hard-coding

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit 5125f8b304b860383f277e6d52bb72abfe820a30)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
[19.03] Dockerfile: move CRIU_VERSION lower

Match the position with where it's on master after the
Dockerfile buildkit refactor.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

cpuguy83 reviewed (approved):
LGTM

tianon commented on Dockerfile:
```diff
@@ -232,52 +251,57 @@ RUN ldconfig
+        zip \
+    && rm -rf /var/lib/apt/lists/*
+
+# Switch to use iptables instead of nftables (to match the host machine)
```
It would be nice if we could somehow detect this and use the appropriate version accordingly instead of just assuming the host uses `iptables` :disappointed:

(maybe a `TODO` for if/when we support nftables?)

thaJeztah replied:
Yeah; this one was mostly taking from what's on master, but I agree that this is pretty much hard coded to the machines we currently have running in our CI

Perhaps you could open a quick PR on master to add the TODO (and a corresponding tracking issue?)

tianon replied:
Oh doh, forgot this was a backport, sorry!  Will do! :+1:

tianon reviewed (approved):
All the `:sort`! :muscle:

thaJeztah commented:
I attached this to the 19.03.8 milestone, as it's not critical for 19.03.7; I suggest we merge this one after 19.03.7 is tagged and done

thaJeztah committed:
Revert "dockerfile: update vndr to 85886e1a"

This reverts commit 0d4f412ecdd3876eb2751d79ec9cf2a2dedd7a15.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Rebased; to get cherry-picks to be as clean as possible, I reverted https://github.com/moby/moby/commit/0d4f412ecdd3876eb2751d79ec9cf2a2dedd7a15, and is re-applied after that. Happy to instead just modify the commits to resolve the conflicts.

PTAL

cpuguy83 reviewed (approved):
LGTM
